src/store/reducers/reducer.js

```diff
@@ -29,20 +35,47 @@ export const reducer = (state, action) => {
       };
     }
     case UPDATE_PRODUCT: {
-      let indexProduct = -1;
-      let finedProduct = false;
-      while (!finedProduct) {
-        indexProduct++;
-        if (action.payload.id === state.product[indexProduct].id) {
-          finedProduct = true;
-        }
-      }
+      const indexProduct = findIndexProductById(
+        action.payload.id,
+        state.product
+      );
       state.product[indexProduct] = action.payload;
       return {
         ...state,
         product: state.product,
       };
     }
+    case ADD_TO_CART: {
+      const { id, quantity } = action.payload;
+      let productToAdd = state.product[findIndexProductById(id, state.product)];
+      let newState = {};
+      productToAdd.isCarted = true;
+
+      const isAdded = () => {
+        return state.cart.find((item) => item.id === id);
+      };
+
+      const changeQuantityInProduct = (product, quantityChoose) => {
+        product.quantity -= quantityChoose;
+      };
+      changeQuantityInProduct(productToAdd, quantity);
+
+      if (!isAdded()) {
+        productToAdd = { ...productToAdd, quantity };
+        newState = {
+          ...state,
+          cart: [...state.cart, productToAdd],
+        };
```

Comment:
just find the index and update the index,don't need to create new state to store the previous state